i18n-helpers/src/bin/mdbook-xgettext.rs

```diff
@@ -86,21 +103,161 @@ fn create_catalog(ctx: &RenderContext) -> anyhow::Result<Catalog> {
+    };
+    catalogs.insert(summary_template_destination, catalog);
+
     // Next, we add the chapter contents.
-    for item in ctx.book.iter() {
-        if let BookItem::Chapter(chapter) = item {
-            let path = match &chapter.path {
+    for item in &ctx.book.sections {
+        if let BookItem::PartTitle(title) = item {
+            // Iterating through the book in section-order, the PartTitle
+            // represents the 'section' that each chapter exists within.
+            current_top_level = slug(title);
+        } else if let BookItem::Chapter(chapter) = item {
+            let source = match &chapter.path {
                 Some(path) => ctx.config.book.src.join(path),
                 None => continue,
             };
+
+            let template_destination = if depth == 0 {
+                PathBuf::from("messages")
+            } else if depth == 1 {
+                PathBuf::from(current_top_level.clone())
+            } else {
+                // The current chapter is already at depth 2, so append the chapter's
+                // name for depths greater than 1.
+                PathBuf::from(current_top_level.clone()).join(slug(&chapter.name))
+            };
+
+            // Add the (destination, catalog) to the map if it doesn't yet
+            // exist, so messages can be appended to the catalog.
+            if let None = catalogs.get(&template_destination) {
+                catalogs.insert(
+                    template_destination.clone(),
+                    Catalog::new(generate_catalog_metadata(&ctx)),
+                );
+            }
+            let catalog: &mut Catalog = catalogs.get_mut(&template_destination).unwrap();
+
             for (lineno, msgid) in extract_messages(&chapter.content) {
-                let source = format!("{}:{}", path.display(), lineno);
-                add_message(&mut catalog, &msgid, &source);
+                let source_with_lineno = format!("{}:{}", source.display(), lineno);
+                add_message(catalog, &msgid, &source_with_lineno);
+            }
+
+            // Add the contents for all of the sub-chapters within the current chapter.
+            for ChapterTemplateInfo {
+                content,
+                source,
+                template_destination,
+            } in get_subcontent_for_chapter(chapter, template_destination, depth, 2)
+            {
+                if let None = catalogs.get(&template_destination) {
+                    catalogs.insert(
+                        template_destination.clone(),
+                        Catalog::new(generate_catalog_metadata(&ctx)),
+                    );
+                }
+                let catalog = catalogs.get_mut(&template_destination).unwrap();
+                let source = ctx.config.book.src.join(&source);
+
+                for (lineno, msgid) in extract_messages(&content) {
+                    let source_with_lineno = format!("{}:{}", source.display(), lineno);
```

Comment:
Would you be okay to use `path` instead of `source` and then rename your `source_with_lineno` back to `source`? That would align with the [`source()` method in polib](https://docs.rs/polib/latest/polib/message/trait.MessageView.html#tymethod.source) (that's where I got my variable names from in the first place).

Reply:
No problem, done!